manifest/infer.go

```diff
@@ -0,0 +1,167 @@
+package manifest
+
+import (
+	"net/http"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/cashapp/hermit/github"
+	"github.com/cashapp/hermit/platform"
+	"github.com/cashapp/hermit/ui"
+	"github.com/pkg/errors"
+)
+
+var (
+	inferVerRe   = regexp.MustCompile(`releases/download/(?:v)?([^/]+)/`)
+	inferOSRe    = regexp.MustCompile(`(darwin|linux)`)
+	inferArchRe  = regexp.MustCompile(`(amd64|arm64)`)
+	inferXArchRe = regexp.MustCompile(`(x86_64|aarch64)`)
+	darwinamd64  = platform.Platform{OS: platform.Darwin, Arch: platform.Amd64}
+	darwinarm64  = platform.Platform{OS: platform.Darwin, Arch: platform.Arm64}
+)
+
+// InferFromArtefact attempts to infer a Manifest from a package artefact.
+//
+// "url" should be the full URL to a package artefact.
+//
+//     https://github.com/protocolbuffers/protobuf-go/releases/download/v1.27.1/protoc-gen-go.v1.27.1.darwin.amd64.tar.gz
+//
+// "version" may be specified if it cannot be inferred from the URL.
+func InferFromArtefact(p *ui.UI, httpClient *http.Client, ghClient *github.Client, url, version string) (*Manifest, error) {
```

Comment:
This function feels a bit long, maybe it could be split up?

Reply:
Yes good idea, I had intended to but forgot!